### PR TITLE
Hide the mutex lock in formatter.ContainerStats

### PR DESCRIPTION
**Pull request by boaz0**
**\- What I did**
Created getters and setters to the container statistics and hide the mutex lock.

**\- How I did it**
- Hide the mutex field in formatter.ContainerStats.
- Expose setters and getters that return or set the statistics using that mutex.

**\- How to verify it**
- Run the integration and unit tests
- Run `docker stats` and `docker stats --format "{{.PIDs}} -- {{.Container}}"` and see that the formats are different.

**\- Description for the changelog**
This patch supposes to fix the issue mentioned in #24987

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @runcom @cpuguy83 PTAL

File: cli/command/container/stats.go [hunk 1]
```diff
@@ -199,12 +207,13 @@ func runStats(dockerCli *command.DockerCli, opts *statsOptions) error {
 	var err error
 	for range time.Tick(500 * time.Millisecond) {
 		cleanHeader()
-		cStats.mu.RLock()
-		csLen := len(cStats.cs)
-		if err = formatter.ContainerStatsWrite(statsCtx, cStats.cs); err != nil {
+		cStats.mutex.RLock()
+		csList := cStats.copyContainersStats()
```
Comment (ripcurld00d) [inline cli/command/container/stats.go]: I hate the idea of copying the containers statistics (it won't scale nicely).
I am working on a better way to avoid this part.

**Comment by ripcurld00d** [comment]
@thaJeztah @runcom @cpuguy83 @vdemeester @LK4D4 
Can you please review this change?

Any feedback will be appreciated! :smiley_cat: 
Thanks a lot!

File: cli/command/container/stats.go (hunk 1 above)
Comment (ripcurld00d) [inline cli/command/container/stats.go]: Fixed

**Comment by LK4D4** [comment]
@ripcurld00d Doesn't this mean that `stats` could get CPU and memory from different time frames? I sorta suspicious about granular locking for setting parts of stats, it adds a lot of code and has little sense IMO, because stats structure supposed to be "atomic" for ps(also repeated lock/unlock might have performance issues). Maybe I'm missing something, though.

**Comment by ripcurld00d** [comment]
@LK4D4, no you got it right. I do believe I can make the read atomic (read the CPU, memory and etc in one time frame), though. 

Anything else that I should keep in mind?

**Comment by LK4D4** [comment]
@ripcurld00d Otherwise, patch looks good to me.

**Comment by ripcurld00d** [comment]
@LK4D4 I updated the commit, feel free to give new feedback.
/cc @runcom 
:grinning:

File: cli/command/formatter/stats.go [hunk 2]
```diff
@@ -37,11 +35,80 @@ type ContainerStatsAttrs struct {
 	PidsCurrent      uint64 // Not used on Windows
 }
 
-// ContainerStats represents the containers statistics data.
+// ClearContainerStats represents the statistics data collected from a container.
+type ClearContainerStats struct {
```
Comment (LK4D4) [inline cli/command/formatter/stats.go]: This name isn't good IMO. At first, it doesn't reflect what this structure does and then "clear containers" is intel containers technology :)

File: cli/command/formatter/stats.go [hunk 3]
```diff
@@ -37,11 +35,80 @@ type ContainerStatsAttrs struct {
...
+
+// SetError sets container statistics error
+func (cs *ContainerStats) SetError(err error) {
+	cs.mutex.Lock()
+	defer cs.mutex.Unlock()
+	cs.err = err
+}
+
+// SetStatistics set the container statistics
+func (cs *ContainerStats) SetStatistics(cpuPerc, mem, memPerc, memLimit, netRx, netTx, blkRead, blkWrite float64, pids uint64) {
```
Comment (LK4D4) [inline cli/command/formatter/stats.go]: Can't you make statistics type public and use it here as argument?

**Review by LK4D4 — CHANGES_REQUESTED** [review]

File: cli/command/formatter/stats.go (hunk 2 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: Yes, I am bad at giving names :blush:.
Is it a bad idea to move this to api/types and call it `ContainerStats`?

File: cli/command/formatter/stats.go (hunk 3 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: Yes, it will definitely make it more readable. :+1:

File: cli/command/formatter/stats.go (hunk 2 above)
Comment (LK4D4) [inline cli/command/formatter/stats.go]: I dunno :) ping @vdemeester @thaJeztah

Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: Anyway, for the time being, I changed it to `ContainerStatsRecord`. Hope it is better than just `ClearContainerStats`.

File: cli/command/formatter/stats.go (hunk 3 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: Fixed

File: cli/command/formatter/stats.go (hunk 2 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: @thaJeztah  `ContainerStats` is used for collecting the container statistics while `ClearContainerStats` (now `ContainerStatsRecord) is used to display the statistics on the screen.

Because of that `ContainerStats` has both mutex and an error field.

**Comment by ripcurld00d** [comment]
Odd, windowRS1 failed:
```
10:59:05 FAIL: docker_cli_run_test.go:1855: DockerSuite.TestRunInteractiveWithRestartPolicy
               docker_cli_run_test.go:1868:
               Failures:
               ExitCode was 0 expected 11
               Expected command to finish, but it hit the timeout

10:18:24 FAIL: docker_cli_events_test.go:81: DockerSuite.TestEventsLimit
               docker_cli_events_test.go:102:
               value *errors.errorString = &errors.errorString{s:"exit status 125: d:\\CI\\CI-7bc23d0\\binary\\docker.exe: Error response from daemon: container f885f96b146b7d711d78d27a0f100c72a5d18fc72a340359c376e1c77f5c1424 encountered an error during Start failed in Win32: This operation returned because the timeout period expired. (0x5b4).\n"} ("exit status 125: d:\\CI\\CI-7bc23d0\\binary\\docker.exe: Error response from daemon: container f885f96b146b7d711d78d27a0f100c72a5d18fc72a340359c376e1c77f5c1424 encountered an error during Start failed in Win32: This operation returned because the timeout period expired. (0x5b4).\n")
```

None of them are related to `docker stats`.
Is my commit the only one who failed on those tests? :worried:

File: cli/command/formatter/stats.go [hunk 4]
```diff
@@ -37,11 +36,72 @@ type ContainerStatsAttrs struct {
 	PidsCurrent      uint64 // Not used on Windows
 }
 
-// ContainerStats represents the containers statistics data.
+// ContainerStatsRecord represents the statistics data collected from a container.
```
Comment (cpuguy83) [inline cli/command/formatter/stats.go]: I think this comment is not right.
`Statistics` is where this comment belongs.

File: cli/command/formatter/stats.go [hunk 5]
```diff
@@ -20,12 +20,11 @@ const (
...
 	pidsHeader       = "PIDS"              // Used only on Linux
+
 )
 
-// ContainerStatsAttrs represents the statistics data collected from a container.
-type ContainerStatsAttrs struct {
-	Windows          bool
-	Name             string
+// Statistics represents the containers resources statistics
+type Statistics struct {
```
Comment (cpuguy83) [inline cli/command/formatter/stats.go]: I think maybe a better name would be `StatsEntry`

File: cli/command/formatter/stats.go (hunk 4 above)
Comment (cpuguy83) [inline cli/command/formatter/stats.go]: Still not quite sure what this does. Maybe it's just for formatting? 
Maybe `ContainerStatsFormatter`?

File: cli/command/formatter/stats.go [hunk 6]
```diff
@@ -37,11 +36,72 @@ type ContainerStatsAttrs struct {
...
-// ContainerStats represents the containers statistics data.
+// ContainerStatsRecord represents the statistics data collected from a container.
+type ContainerStatsRecord struct {
+	name     string
+	windows  bool
+	notValid bool
+	Statistics
+}
+
+// ContainerStats represents the container statistics for collecting.
```
Comment (cpuguy83) [inline cli/command/formatter/stats.go]: All of these with basically the same fields and name is pretty confusing.

File: cli/command/formatter/stats.go [hunk 7]
```diff
@@ -37,11 +36,72 @@ type ContainerStatsAttrs struct {
...
+
+// SetStatistics set the container statistics
+func (cs *ContainerStats) SetStatistics(s *Statistics) {
+	cs.mutex.Lock()
+	defer cs.mutex.Unlock()
+	cs.Statistics = *s
+}
+
+// GetStatistics returns container statistics with other meta data such as the container name
+func (cs *ContainerStats) GetStatistics() ContainerStatsRecord {
```
Comment (cpuguy83) [inline cli/command/formatter/stats.go]: Not really "getting" anything here, just copying into a new struct.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
I can sort of follow the code, but it's a bit confusing.

I wonder if we need so many different structs for the same information.
Instead we can put methods on one, like `IsWindows()` which can lock and check the `windows` field, and `Name()` and read the `name` field.

File: cli/command/formatter/stats.go (hunk 5 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: :+1:

File: cli/command/formatter/stats.go (hunk 7 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: Yes, that's right. Nevertheless, you eventually get a `ContainerStatsRecord` entity.
You want me to rename this anyway?

File: cli/command/formatter/stats.go (hunk 4 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: :+1: sounds right

**Comment by ripcurld00d** [comment]
@cpuguy83 thanks for your feedback.
I really appreciate your opinion just need some clarification.

Thanks.

File: cli/command/formatter/stats.go (hunk 7 above)
Comment (cpuguy83) [inline cli/command/formatter/stats.go]: Doesn't matter.

**Comment by ripcurld00d** [comment]
@cpuguy83 and @LK4D4 PTAL thanks. :smiley_cat: 

/cc @thaJeztah

File: cli/command/formatter/stats.go [hunk 8]
```diff
@@ -35,19 +36,74 @@ type ContainerStatsAttrs struct {
...
+	cs.mutex.Lock()
+	defer cs.mutex.Unlock()
+	cs.err = err
+	if err != nil {
+		cs.IsInvalid = true
+	}
+}
+
+// SetStatistics set the container statistics
+func (cs *ContainerStats) SetStatistics(s *StatsEntry) {
```
Comment (LK4D4) [inline cli/command/formatter/stats.go]: Let's use values instead of pointers everywhere for consistency.

File: cli/command/formatter/stats.go [hunk 9]
```diff
@@ -35,19 +36,74 @@ type ContainerStatsAttrs struct {
...
+func (cs *ContainerStats) GetStatistics() StatsEntry {
+	cs.mutex.Lock()
+	defer cs.mutex.Unlock()
+	return cs.StatsEntry
 }
 
 // NewStatsFormat returns a format for rendering an CStatsContext
 func NewStatsFormat(source, osType string) Format {
+	if osType == "windows" {
+		isWindowsDaemon = true
```
Comment (LK4D4) [inline cli/command/formatter/stats.go]: I don't like global variable usage here. It's definitely shouldn't be global.

**Review by LK4D4 — CHANGES_REQUESTED** [review]

File: cli/command/formatter/stats.go (hunk 9 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: Ummm... the alternative is to put this variable inside the StatsEntity struct.
What do you think?

Comment (LK4D4) [inline cli/command/formatter/stats.go]: Yeah, I think it would be better. I saw that we use global var in stat_helper too. It's also so wrong :(

File: cli/command/formatter/stats.go (hunk 8 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: :+1: no problem

File: cli/command/formatter/stats.go (hunk 9 above)
Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: :+1: 
I don't mind fixing this as well.
The question is whether you prefer doing that in this commit or in another one.

Comment (LK4D4) [inline cli/command/formatter/stats.go]: Let's do it in another PR.

Comment (ripcurld00d) [inline cli/command/formatter/stats.go]: no problem :coffee:

**Comment by ripcurld00d** [comment]
@LK4D4 re-review :8ball: :sake: plz?

**Review by LK4D4 — APPROVED** [review]

**Comment by LK4D4** [comment]
ping @cpuguy83

**Comment by tonistiigi** [comment]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM
